lib/sensors/src/URM04Sensor.cpp

```diff
@@ -0,0 +1,216 @@
+    // no error occured in trigger step
+    distance = 0;
+  }
+
+  // flush the buffer from serial
+  serial.sync();
+
+  // wait for at least 30 ms after calling trigger function
+  ThisThread::sleep_for(35ms);
+
+  // reset command buffer - fill whole array with zeros
+  memset(&cmdst[0], 0, sizeof(cmdst));
+}
+
+// pass by reference a variable to hold the distance value - will give MAX_FLOAT if out of range
+bool URM04Sensor::URM04Sensor::read_distance(float& distance) {
+  /****************** TRIGGER SENSOR BEFORE READING DISTANCE ********************/
+  trigger_sensor(distance);
+
+  /************ INSTANTIATE COMMANDS TO BE SENT OVER SERIAL************/
+  // check sum represents the final bit in command buffer - made by adding all previous bits in command buffer
+  uint8_t checkSum;
+  // buffer header
+  cmdst[0] = 0x55;
+  cmdst[1] = 0xAA;
+  // device address
+  cmdst[2] = startAddr;
+  // command length
+  cmdst[3] = 0x00;
+  // the command itself
+  cmdst[4] = 0x02;
+  // checksum bit
+  checkSum = cmdst[0] + cmdst[1] + cmdst[2] + cmdst[3] + cmdst[4];
+  // instantiate the last element in the command buffer with checksum value
+  cmdst[5] = checkSum;
+
+  /************** SEND COMMANDS OVER SERIAL***********/
+  int w_num_bytes;
+  // returns the number of bytes if successful read
+  w_num_bytes = serial.write(&cmdst[0], sizeof(cmdst));
+
+  // else if numb_btyes < 0 serial failed or if number of bytes read != 6 (because 6 bytes sent over serial)
+  if (w_num_bytes < 0 || w_num_bytes != 6) {
+    // return false indicating read command has failed
+    distance = -1;
+    return false;
+  }
+
+  // flush buffer from serial
+  serial.sync();
+
+  // reset command buffer - fill whole array with zeros
+  memset(&cmdst[0], 0, sizeof(cmdst));
+  /******* READ RETURN VALUE FROM SERIAL**************/
+
+  // turn on reading mode for RS485
+  m_trigPin.write(LOW);
+  ThisThread::sleep_for(1ms);
+
+  int r_num_bytes;
+  // returns the number of bytes if successful read
+  r_num_bytes = serial.read(&cmdst[0], sizeof(cmdst));
+
+  // else if numb_btyes < 0 serial failed or if number of bytes read != 8 (because 8 bytes sent over serial)
+  if (r_num_bytes < 0 || r_num_bytes != 8) {
+    distance = -1;
+    return false;
+  }
+  /******** PARSE THROUGH THE DATA READ FROM SERIAL*********/
+  else {
+    // create the checksum
+    checkSum = cmdst[0] + cmdst[1] + cmdst[2] + cmdst[3] + cmdst[4] + cmdst[5] + cmdst[6];
+
+    // check if the checksum is incorrect
+    if (checkSum != cmdst[7]) {
+      // return false if checksum failed
+      distance = -1;
+      return false;
+    } else {
+      // get distance from sensor
+      distance = (float)(cmdst[5] << 8) + (float)cmdst[6];
+      // --------------------- or --------------------------
+      // distance = (float)(cmdst[5]*256) + (float)cmdst[6];
+
+      // check if distance is out of range - if so return maximum float value
+      if ((int)distance % (int)cmdst[5] == (int)cmdst[6]) {
```

Comment:
I don't understand the logic here, the URM04 guide says it'll return 0xFFFF if the distance is out of range, so just do `if (cmdst[5] == 0xFF && cmdst[6] == 0xFF)`?